v2g-liberty/rootfs/root/appdaemon/apps/v2g_liberty/modbus_evse_client.py

```diff
@@ -803,80 +838,131 @@ async def __get_and_process_registers(self, entities: list):
+        :return: Nothing
+        """
+        current_value = evse_entity["current_value"]
+
+        if current_value == new_value:
+            # self.__log(f"{evse_entity['ha_entity_name']} {new_value=} is same as current, ignore.")
+            pass
+        else:
+            evse_entity["current_value"] = new_value
+            evse_entity["previous_value"] = current_value
+
+            # Call change_handler if defined
+            if "change_handler" in evse_entity.keys():
+                str_action = evse_entity["change_handler"]
                 # TODO: Find an more elegant way (without 'eval') to do this...
                 if str_action == "__handle_charger_state_change":
-                    await self.__handle_charger_state_change(new_charger_state=value)
+                    await self.__handle_charger_state_change(
+                        new_charger_state=new_value,
+                        old_charger_state=current_value,
+                    )
+                elif str_action == "__handle_charger_error_state_change":
+                    # This is the case for the ENTITY_ERROR_1..4. The charger_state
+                    # does not necessarily change only (one or more of) these error-states.
+                    # So the state is not added to the call.
+                    await self.__handle_charger_error_state_change({"dummy": None})
                 else:
-                    self.__log(f"__update_entity unknown action: '{str_action}'.")
+                    self.__log(f"unknown action: '{str_action}'.", level="WARNING")
 
-    async def __update_state(self, entity_id, state=None, attributes=None):
-        """Generic function for updating the state of an entity in Home Assistant
-            If it does not exist, create it.
-            If it has attributes, keep them (if not overwrite with empty)
+    async def __update_ha_entity(
+        self, entity_id: str, new_value=None, attributes: dict = {}
+    ):
+        """
+        Generic function for updating the state of an entity in Home Assistant
+        If it does not exist, create it.
+        If it has attributes, keep them (if not overwrite with empty)
 
         Args:
             entity_id (str): full entity_id incl. type, e.g. sensor.charger_state
-            state (any, optional): The value the entity should be written with. Defaults to None.
-            attributes (any, optional): The value (can be dict) the attributes should be written
-            with. Defaults to None.
+            new_value (any, optional): The value the entity should be written with.
+              Defaults to None, can be "unknown".
+            attributes (dict, optional): The dict the attributes should be written
+            with. Defaults to {}.
         """
-        if state is None:
-            self.__log("called with state is None, aborting.")
-            return
-
-        new_attributes = None
+        new_attributes = {}
```

Comment:
This function is a lot more complicated than strictly needed by the handling of the attributes. This is used only once. Maybe we are better of with a special function for that single use-case?

Reply:
I would expect that set_state does that by default. If that is not the case, then I guess this will have to do.